pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,219 @@
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+// it will only remove the hook once it has identified that the Machine resource is being deleted and it doesn't host an etcd member
+func (c *machineDeletionHooksController) attemptToDeleteMachineDeletionHook(ctx context.Context, recorder events.Recorder) error {
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	machinesWithHooks := ceohelpers.FilterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if at least one member machine is pending deletion
+	machinesPendingDeletion := ceohelpers.FilterMachinesPendingDeletion(machinesWithHooks)
+	if len(machinesPendingDeletion) == 0 {
+		klog.V(4).Infof("currently, we don't have a machine pending deletion")
+		return nil
+	}
+
+	// get the members, issue a live call instead of reading from the config map
+	// because we need to take into account learners
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+	currentMemberIPListSet := sets.NewString()
+	for _, member := range members {
+		memberIP, err := ceohelpers.MemberToNodeInternalIP(member)
+		if err != nil {
+			return err
+		}
+		currentMemberIPListSet.Insert(memberIP)
+	}
+
+	// find a machine that doesn't have a member
+	for machinePendingDeletionIP, machinePendingDeletion := range indexMachinesByNodeInternalIP(machinesPendingDeletion) {
+		if currentMemberIPListSet.Has(machinePendingDeletionIP) {
+			// Ignore machines that still have a member present
+			continue
+		}
+		machinePendingDeletionCopy := machinePendingDeletion.DeepCopy()
+		removeMachineDeletionHook(machinePendingDeletionCopy)
+		if _, err := c.machineClient.Update(ctx, machinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
+			klog.V(2).Infof("successfully removed the deletion hook from machine %v", machinePendingDeletion.Name)
+			recorder.Eventf("MachineDeletionHook", "successfully removed the deletion hook from machine %v as it doesn't have a member", machinePendingDeletion.Name)
+		}
+		return err
+	}
+	return nil
+}
+
+// addDeletionHookToMasterMachines adds the Machine Deletion Hook to master machines
+func (c *machineDeletionHooksController) addDeletionHookToMasterMachines(ctx context.Context) error {
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	for _, masterMachine := range masterMachines {
+		if masterMachine.DeletionTimestamp != nil {
+			continue
+		}
+		if hasCompleteMachineDeletionHook(masterMachine) {
+			continue
+		}
+		masterMachineCopy := masterMachine.DeepCopy()
+		if hasPartialMachineDeletionHook(masterMachineCopy) {
+			repairPartialMachineDeletionHook(masterMachineCopy)
+		} else {
+			addPreDrainHook(masterMachineCopy)
+		}
+		if _, err := c.machineClient.Update(ctx, masterMachineCopy, metav1.UpdateOptions{}); err != nil {
+			return err
```

Comment:
aggregate the errors from all so that you can make partial progress and still fail.